speechbrain/core.py

```diff
@@ -1151,6 +1165,24 @@ def _fit_train(self, train_set, epoch, enable):
         self.step = 0
         self.valid_step = 0
 
+    def _should_save_intra_epoch_ckpt(self, last_ckpt_time, steps_since_ckpt):
+        """Determines if an intra-epoch checkpoint should be saved.
+
+        Returns True if there's a checkpointer and time or steps has exceeded limit.
+        """
+        if self.checkpointer is None:
+            return False
+
+        # Check if we've run for the requested amount of time
+        if 0 < self.ckpt_interval_minutes * 60.0 < time.time() - last_ckpt_time:
+            # Only save on the main process to avoid race conditions
+            return if_main_process()
+
+        # Save after requested # of steps. This option is the only one that
+        # allows saving on multiple processes. The logic for whether saving
+        # is run only on the main process is handled by the checkpointer.
+        return 0 < self.ckpt_interval_steps <= steps_since_ckpt
+
```

Comment:
With the separate early returns, I think a tricky bug could happen if someone uses both `ckpt_interval_minutes` and `ckpt_interval_steps`. There could be cases where the time interval and step interval happen simultaneously.

I suggest:
```suggestion
        should_save = False
        # Check if we've run for the requested amount of time
        if 0 < self.ckpt_interval_minutes * 60.0 < time.time() - last_ckpt_time:
            # Only save on the main process to avoid race conditions
            should_save = if_main_process()

        # Save after requested # of steps. This option is the only one that
        # allows saving on multiple processes. The logic for whether saving
        # is run only on the main process is handled by the checkpointer.
        if 0 < self.ckpt_interval_steps <= steps_since_ckpt:
            should_save = True
        
        return should_save

```

Reply:
Actually now that I think about it, resetting both steps since last checkpoint and the time since last checkpoint will make the main process quickly out of sync when both methods are used. Should we make steps and time exclusive? Or maybe just run two different checks in the loop?